src/graphapi/matmul.cpp

```diff
@@ -104,5 +105,310 @@ void BackendMatmulDescriptor::getAttribute(miopenBackendAttributeName_t attribut
+{
+    mOperationMatmul.mA = checkPtr(A);
+    if(mOperationMatmul.mA->getDimensions().size() < 2)
+        MIOPEN_THROW(miopenStatusBadParm);
+    aSet = true;
+    return *this;
+};
+OperationMatmulBuilder& OperationMatmulBuilder::setB(Tensor* B)
+{
+    mOperationMatmul.mB = checkPtr(B);
+    if(mOperationMatmul.mB->getDimensions().size() < 2)
+        MIOPEN_THROW(miopenStatusBadParm);
+    bSet = true;
+    return *this;
+};
+OperationMatmulBuilder& OperationMatmulBuilder::setC(Tensor* C)
+{
+    mOperationMatmul.mC = checkPtr(C);
+    if(mOperationMatmul.mC->getDimensions().size() < 2)
+        MIOPEN_THROW(miopenStatusBadParm);
+    cSet = true;
+    return *this;
+};
+OperationMatmulBuilder& OperationMatmulBuilder::setBatchCount(int64_t count)
+{
+    mOperationMatmul.mBatchCount = count;
+    return *this;
+};
+OperationMatmulBuilder& OperationMatmulBuilder::setGemmMOverride(Tensor* overrideTensor)
+{
+    mOperationMatmul.mGemmMOverride = checkPtr(overrideTensor);
+    return *this;
+};
+OperationMatmulBuilder& OperationMatmulBuilder::setGemmNOverride(Tensor* overrideTensor)
+{
+    mOperationMatmul.mGemmNOverride = checkPtr(overrideTensor);
+    return *this;
+};
+OperationMatmulBuilder& OperationMatmulBuilder::setGemmKOverride(Tensor* overrideTensor)
+{
+    mOperationMatmul.mGemmKOverride = checkPtr(overrideTensor);
+    return *this;
+};
+OperationMatmulBuilder& OperationMatmulBuilder::setMatmulDescriptor(Matmul* mMatmul)
+{
+    mOperationMatmul.mMatmul = checkPtr(mMatmul);
+    matmulSet                = true;
+    return *this;
+}
+
+OperationMatmul OperationMatmulBuilder::build()
+{
+    if(!aSet || !bSet || !cSet || !matmulSet)
+        MIOPEN_THROW(miopenStatusBadParm);
+
+    int aDimensionsCount = mOperationMatmul.mA->getDimensions().size();
+    int bDimensionsCount = mOperationMatmul.mB->getDimensions().size();
+    int cDimensionsCount = mOperationMatmul.mC->getDimensions().size();
+
+    if(aDimensionsCount != bDimensionsCount || bDimensionsCount != cDimensionsCount)
+        MIOPEN_THROW(miopenStatusBadParm);
+
+    int Am = mOperationMatmul.mA->getDimensions().end()[-2];
+    int An = mOperationMatmul.mA->getDimensions().end()[-1];
+
+    int Bn = mOperationMatmul.mB->getDimensions().end()[-2];
+    int Bk = mOperationMatmul.mB->getDimensions().end()[-1];
+
+    int Cm = mOperationMatmul.mB->getDimensions().end()[-2];
+    int Ck = mOperationMatmul.mB->getDimensions().end()[-1];
+
+    if(Am != Cm || An != Bn || Bk != Ck)
+        MIOPEN_THROW(miopenStatusBadParm);
+
+    // TODO: need broadcsat checks
+    //
+
+    return mOperationMatmul;
+}
+
+void BackendOperationMatmulDescriptor::finalize()
+{
+    if(mFinalized || !deref(mA).isFinalized() || !deref(mB).isFinalized() ||
+       !deref(mC).isFinalized() || !deref(mGemmMOverride).isFinalized() ||
+       !deref(mGemmNOverride).isFinalized() || !deref(mGemmKOverride).isFinalized())
+    {
+        MIOPEN_THROW(miopenStatusNotInitialized);
+    }
```

Comment:
For invalid attributes we throw `miopenStatusBadParm`.

By the time of execution of this function the fulfillment of `isFinalized()` requirement was already checked in `setAttribute()`, I believe, there is no need of redoing it here.

At first I wanted to omit `isFinalized()` checks:

```suggestion
    if(mFinalized)
    {
        MIOPEN_THROW(miopenStatusNotInitialized);
    }
```

but upon reviewing cuDNN's docs [cudnnbBackendFinalize()](https://docs.nvidia.com/deeplearning/cudnn/latest/api/cudnn-graph-library.html#cudnnbackendfinalize) once again I think we need to simply return here:

```suggestion
    if(mFinalized)
    {
        return;
    }
```

So I was wrong in `throw`ing in all such cases in my PRs. Could you also take a look at the cuDNN documentation and give your opinion, please?

Reply:
It's check for double finalization

Reply:
It should return CUDNN_STATUS_BAD_PARAMS

Reply:
> It should return CUDNN_STATUS_BAD_PARAMS

User can check result of the function call via std::optional<>, status code, catching exception or input-output parameters. At least one way should exist. So, throwing exception is better then return void, but returning status code does not call std::terminate() in worst case scenario. Crashing application from library is bad idea

Reply:
@kikimych 
> So, throwing exception is better then return void, but returning status code does not call std::terminate() in worst case scenario.

~The initial idea was to format all the user's hard disks, but finally~ we ~decided to~ catch exceptions at API wrapper level and return error code. Please check yourself.

Reply:
got it